### ISSUE TITLE
Debuggability: Wrong type in argument matchers results in NullReferenceException

### ISSUE THREAD

**Issue by TimLovellSmith** [FakeItEasy/FakeItEasy#660]
(With FakeItEasy 2.0.0-beta009)
I have a test like below which passes.

If I accidentaly changed `A<DateTimeOffset?>` to `A<DateTimeOffset>` (non nullable type) I get a rather tricky to debug nullref:

> System.NullReferenceException occurred
>   _HResult=-2147467261
>   _message=Object reference not set to an instance of an object.
>   HResult=-2147467261
>   IsTransient=false
>   Message=Object reference not set to an instance of an object.
>   Source=FakeItEasy
>   StackTrace:
>        at FakeItEasy.Core.DefaultArgumentConstraintManager`1.MatchesConstraint.FakeItEasy.Core.IArgumentConstraint.IsValid(Object argument)
>   InnerException: 

``` c#
var provisioner = CreateProvisioner();
var operations = CreateOperations();

Deployment dep = new Deployment();
A.CallTo(() => provisioner.GetResource(A<ResourceSpec>._, A<DateTimeOffset?>._))
    .Returns(dep);

var deleteResourceCall = A.CallTo(() => operations.DeleteDeployment(dep, A<RequestData>._));

await WithTestServer(operations, provisioner)(async (server) =>
{
    string uri = CreateUri();
    var response = await server.HttpClient.DeleteAsync(uri);

    Assert.Equal(HttpStatusCode.OK, response.StatusCode);
});

deleteResourceCall.MustHaveHappened();
```

**Comment by thomaslevesque** [comment]
Thanks for the report @TimLovellSmith ! I have reproduced the bug; basically FakeItEasy is trying to cast null to `DateTimeOffset`, and of course it fails.

The most obvious fix would be to say that `null` never matches a value-type constraint. In other words, if you pass `A<DateTimeOffset>._` when you configure the call, and then call the method with null, it won't match.

**Comment by adamralph** [comment]
Agreed.

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 3.1.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/3.1.0).

@TimLovellSmith, look for your name in the release notes! :trophy:
